### PR TITLE
schema/gui/generic - use f_color

### PR DESCRIPTION
**Pull request by white-haired-uncle**
Updated to match canvas_private.hpp

Also noticed this on line 94, for drawing object [line]

`{DEFAULT_KEY "alpha" f_unsigned 255}`

couldn't find any reference to this in the code or wiki?

### CODE REVIEW

File: data/schema/gui/generic.cfg [hunk 1]
```diff
@@ -146,7 +146,7 @@
                 {DEFAULT_KEY "text_alignment" f_h_align "left"}
                 {DEFAULT_KEY "text_markup" f_bool false}
                 {DEFAULT_KEY "text_link_aware" f_bool false}
-                {DEFAULT_KEY "text_link_color" string "#ffff00"}
+                {DEFAULT_KEY "text_link_color" f_color "#ffff00"}
```
Comment (soliton-) [inline data/schema/gui/generic.cfg:149]: What about `highlight_color` a few lines up?

Comment (white-haired-uncle) [reply data/schema/gui/generic.cfg:149]: yes, thanks

**Comment by soliton-** [comment]
There is also:
> data/schema/gui/widget_definitions.cfg:        {DEFAULT_KEY "link_color" string "#ffff00"}
data/schema/gui/widget_definitions.cfg:        {DEFAULT_KEY "link_color" string "#ffff00"}
data/schema/gui/generic.cfg:        {DEFAULT_KEY "debug_border_color" color ""}

**Comment by white-haired-uncle** [comment]
> data/schema/gui/widget_definitions.cfg

I looked at debug_border, but I'm pretty sure it's a color, not f_color.

I believe the others are as well?

```label_definition::resolution::resolution(const config& cfg)
    : resolution_definition(cfg)
    , link_color(cfg["link_color"].empty() ? color_t::from_hex_string("ffff00") : color_t::from_rgba_string(cfg["link_color"].str()))
```

**Comment by soliton-** [comment]
Btw, `color` is defined as three comma separated numbers but `from_rgba_string` allows an optional fourth number for the alpha value. That sounds like the schema should also allow 4 numbers? I guess other `color` uses correspond to `from_rgb_string` where it's strictly three numbers?

**Comment by white-haired-uncle** [comment]
Looking at data/schema/types/basic.cfg
```
[type]
    name=color
    [list]
        min=3
        max=4
        [element]
            value="(?:2[0-5][0-5]|[01]?\d?\d)\s*"
        [/element]
    [/list]
[/type]
```
and assuming that's even the right place to look, while I can't fully parse that regex (pretty sure it says 0-255 with optional trailing whitespace, though '?:' eludes me), the min3/max4 makes me think color is defined as either RGB or RGBA?

BTW, https://wiki.wesnoth.org/GUIVariable#color documents the type color as RGBA, not sure if this should be updated (I changed the alpha definition the other day, but not the 4 value aspect).

**Comment by white-haired-uncle** [comment]
Actually, is that regex correct?  I don't see how you could express something like 228 (or anything 2xy, where 0<=x<=4 and 6<=y<=9).

**Comment by soliton-** [comment]
Oh, I see. I looked at https://github.com/wesnoth/wesnoth/blob/e460d26dfc2adda2bb2a4988b08b36e198632410/data/schema/game_config.cfg#L251-L254

but that is indeed not the one used for the GUI schema. Confusing use of the same type name for different things...

And that is certainly a weird regex. 206 is not a valid value?

**Comment by soliton-** [comment]
> Actually, is that regex correct? I don't see how you could express something like 228 (or anything 2xy, where 0<=x<=4 and 6<=y<=9).

Yeah. Something like `value="(?:25[0-5]|2[0-4][0-9]|[01]?\d?\d)\s*"` should work.

**Comment by white-haired-uncle** [comment]
I'm assuming this is correct (I came up with the same), though I don't know how to test.  I think I once saw values being validated and not just keys, but I have no idea where and I don't think it happens as often as I'd like it to.

**Comment by soliton-** [comment]
Validate a config with a color attribute? I don't know what keys-only validation you mean.

GUI config is validated every start automatically. If you edit `data/gui/themes/default/widget/listbox_default.cfg` and change the 255 in `border_color` to 206 for example then validation fails without your PR.

File: data/schema/types/basic.cfg [hunk 2]
```diff
@@ -44,7 +44,7 @@
 		min=3
 		max=4
 		[element]
-			value="(?:2[0-5][0-5]|[01]?\d?\d)\s*"
+            value="(?:25[0-5]|2[0-4][0-9]|[01]?\d?\d)\s*"
```
Comment (soliton-) [inline data/schema/types/basic.cfg:47]: You changed the tabs to spaces. Please leave the indentation as is.

Comment (white-haired-uncle) [reply data/schema/types/basic.cfg:47]: hmm, i just cut and paste.  looks like my editor is changing NEW tabs to spaces

**Comment by white-haired-uncle** [comment]
> Validate a config with a color attribute? I don't know what keys-only validation you mean.
> 
> GUI config is validated every start automatically. If you edit `data/gui/themes/default/widget/listbox_default.cfg` and change the 255 in `border_color` to 206 for example then validation fails without your PR.

I think I've seen values validated, once maybe, but I have no idea where/why/how.

I just edited label_default.cfg:

```
        #link_color = "255, 225, 0"
        link_color = "455, 225, 0"
        link_colour = "255, 226, 0"
```
and I get errors about link_colour= being invalid, but noting about link_color being out of range.

**Comment by white-haired-uncle** [comment]
> Something is very wrong here. These attributes specify a default value that doesn't match the regex for the colour type. Is the `#xxxxxx` syntax accepted by these keys? If so, it should be included in the regex. If not, it shouldn't be used for the default values.

I wouldn't think so...

```
label_definition::resolution::resolution(const config& cfg)
    : resolution_definition(cfg)
    , link_color(cfg["link_color"].empty() ? color_t::from_hex_string("ffff00") : color_t::from_rgba_string(cfg["link_color"].str()))

```
but in trying to test it, not only can I not seem to change the link color regardless of what I put in the label_definition (and yes, it is link_aware while I'm testing), the link is white, not yellow.  Pretty sure it defaulted to yellow the other day when I was testing setting link_color via Lua.

BTW, just what does the default value in the schema mean anyway?  Does it actually do anything?  If it did actually supply the default value, the above wouldn't seem to make much sense (aside from playing it safe?) as cfg["link_color"] shouldn't ever be empty?

**Review by CelticMinstrel — COMMENTED** [review]
> BTW, just what does the default value in the schema mean anyway?  Does it actually do anything?  If it did actually supply the default value, the above wouldn't seem to make much sense (aside from playing it safe?) as cfg["link_color"] shouldn't ever be empty? 

It doesn't currently do anything, other than provide information to people perusing the schema. Even so, it really should be expressed in the format specified by the schema. The schema schema validation should probably validate the default value of a tag against its type. (Yes, I said "schema" twice there, that isn't an error.)

**Comment by white-haired-uncle** [comment]
not sure about the f_colors that default to `#abcdef`

**Review by CelticMinstrel — COMMENTED** [review]
The defaults should be a comma-separated triple of decimal numbers, yes.

**Review by soliton- — APPROVED** [review]

**Review by CelticMinstrel — COMMENTED** [review]
Pretty sure there are more cases in the schema that specify a colour default incorrectly. Should we change those too in this PR or is it better to handle that separately?

**Comment by white-haired-uncle** [comment]
Looking at `grep -r color data/schema` I only see six other "colors" that specify a default, all canvas shapes and all defaults are "".  Four of those have no default in the initializer list, and two have  color_t::null_color().

**Pull request merged**